File: python/vibedds/qos.py
"""QoS policies and RxO (Requested/Offered) compatibility matching."""

from __future__ import annotations

import struct
from dataclasses import dataclass
from enum import IntEnum

from vibedds.cdr import CdrSerializer
from vibedds.types import Duration


class ReliabilityKind(IntEnum):
    BEST_EFFORT = 1
    RELIABLE = 2


class DurabilityKind(IntEnum):
    VOLATILE = 0
    TRANSIENT_LOCAL = 1
    TRANSIENT = 2
    PERSISTENT = 3


class OwnershipKind(IntEnum):
    SHARED = 0
    EXCLUSIVE = 1


class LivelinessKind(IntEnum):
    AUTOMATIC = 0
    MANUAL_BY_PARTICIPANT = 1
    MANUAL_BY_TOPIC = 2


class DestinationOrderKind(IntEnum):
    BY_RECEPTION_TIMESTAMP = 0
    BY_SOURCE_TIMESTAMP = 1


class HistoryKind(IntEnum):
    KEEP_LAST = 0
    KEEP_ALL = 1


class DataRepresentationId(IntEnum):
    XCDR1 = 0
    XML = 1
    XCDR2 = 2


class TypeConsistencyKind(IntEnum):
    DISALLOW_TYPE_COERCION = 0
    ALLOW_TYPE_COERCION = 1


@dataclass
class QosPolicy:
    """Aggregated QoS settings for an endpoint."""
    reliability: ReliabilityKind = ReliabilityKind.BEST_EFFORT
    reliability_max_blocking_time: Duration = Duration.ZERO
    durability: DurabilityKind = DurabilityKind.VOLATILE
    ownership: OwnershipKind = OwnershipKind.SHARED
    liveliness: LivelinessKind = LivelinessKind.AUTOMATIC
    liveliness_lease_duration: Duration = Duration.INFINITE
    deadline_period: Duration = Duration.INFINITE
    destination_order: DestinationOrderKind = DestinationOrderKind.BY_RECEPTION_TIMESTAMP
    history: HistoryKind = HistoryKind.KEEP_LAST
    history_depth: int = 1
    partition: list[str] | None = None


# Default QoS for builtin endpoints (SEDP)
BUILTIN_ENDPOINT_QOS = QosPolicy(
    reliability=ReliabilityKind.RELIABLE,
    durability=DurabilityKind.TRANSIENT_LOCAL,
    history=HistoryKind.KEEP_LAST,
    history_depth=1,
)


def qos_compatible(offered: QosPolicy, requested: QosPolicy) -> bool:
    """Check if offered QoS is compatible with requested QoS (RxO rules).

    Returns True if the writer's offered QoS satisfies the reader's requested QoS.
    """
    # Reliability: offered must be >= requested
    if offered.reliability < requested.reliability:
        return False

    # Durability: offered must be >= requested
    if offered.durability < requested.durability:
        return False

    # Ownership: must match exactly
    if offered.ownership != requested.ownership:
        return False

    # Liveliness: offered must be >= requested (stricter is higher)
    if offered.liveliness < requested.liveliness:
        return False

    # Destination order: offered must be >= requested
    if offered.destination_order < requested.destination_order:
        return False

    # Deadline: offered period must be <= requested period
    # (shorter deadline satisfies longer requirement)
    # We skip this for now since Duration comparison isn't trivial

    # Partition: if both specify partitions, they must have at least one in common
    # Empty / None means default partition which matches default partition
    if offered.partition is not None and requested.partition is not None:
        if not set(offered.partition) & set(requested.partition):
            return False

    return True


def serialize_reliability_qos(qos: QosPolicy, endian: str = "<") -> bytes:
    """Serialize reliability QoS for SEDP parameter list."""
    # kind(u32) + max_blocking_time(Duration: i32+u32) = 12 bytes
    kind = int(qos.reliability)
    return struct.pack(
        endian + "IiI",
        kind,
        qos.reliability_max_blocking_time.seconds,
        qos.reliability_max_blocking_time.fraction,
    )


def deserialize_reliability_qos(data: bytes, endian: str = "<") -> tuple[ReliabilityKind, Duration]:
    """Deserialize reliability QoS from parameter value."""
    kind, secs, frac = struct.unpack(endian + "IiI", data[:12])
    return ReliabilityKind(kind), Duration(seconds=secs, fraction=frac)


def serialize_durability_qos(qos: QosPolicy, endian: str = "<") -> bytes:
    """Serialize durability QoS for SEDP parameter list."""
    return struct.pack(endian + "I", int(qos.durability))


def deserialize_durability_qos(data: bytes, endian: str = "<") -> DurabilityKind:
    """Deserialize durability QoS from parameter value."""
    kind = struct.unpack(endian + "I", data[:4])[0]
    return DurabilityKind(kind)


def serialize_partition_qos(qos: QosPolicy, endian: str = "<") -> bytes:
    """Serialize PartitionQosPolicy (sequence<string>) in XCDR1."""
    ser = CdrSerializer(endian)
    ser.set_origin()
    if qos.partition:
        ser.write_uint32(len(qos.partition))
        for part in qos.partition:
            ser.write_string(part)
    else:
        ser.write_uint32(0)
    return ser.getvalue()


def serialize_data_representation_qos(
    representations: list[DataRepresentationId],
    endian: str = "<",
) -> bytes:
    """Serialize DataRepresentationQosPolicy (sequence<short>) in XCDR1."""
    ser = CdrSerializer(endian)
    ser.set_origin()
    ser.write_sequence(
        representations,
        lambda s, rep: s.write_int16(int(rep)),
    )
    return ser.getvalue()


RTI_DATA_REPRESENTATION_DEFAULT = bytes.fromhex(
    "010000000000000007000000"
)
RTI_TYPE_CONSISTENCY_DEFAULT = bytes.fromhex(
    "0100010100000000"
)


def serialize_data_representation_qos_rti(
    representations: list[DataRepresentationId],
    endian: str = "<",
    tail: int | None = None,
) -> bytes:
    """Serialize DataRepresentationQosPolicy using RTI-observed layout.

    RTI observed a 12-byte payload for HelloWorld:
      01 00 00 00 00 00 00 00 07 00 00 00
    The exact semantics are unclear, so we allow an optional tail value.
    """
    if representations == [DataRepresentationId.XCDR1] and tail is None:
        return RTI_DATA_REPRESENTATION_DEFAULT
    count = len(representations)
    data = bytearray()
    data.extend(struct.pack(endian + "I", count))
    for rep in representations:
        data.extend(struct.pack(endian + "I", int(rep)))
    if tail is not None:
        data.extend(struct.pack(endian + "I", tail))
    return bytes(data)


def serialize_type_consistency_enforcement_qos(
    kind: TypeConsistencyKind = TypeConsistencyKind.DISALLOW_TYPE_COERCION,
    ignore_sequence_bounds: bool = False,
    ignore_string_bounds: bool = False,
    ignore_member_names: bool = False,
    prevent_type_widening: bool = False,
    force_type_validation: bool = False,
    endian: str = "<",
) -> bytes:
    """Serialize TypeConsistencyEnforcementQosPolicy in XCDR1."""
    ser = CdrSerializer(endian)
    ser.set_origin()
    ser.write_int32(int(kind))
    ser.write_bool(ignore_sequence_bounds)
    ser.write_bool(ignore_string_bounds)
    ser.write_bool(ignore_member_names)
    ser.write_bool(prevent_type_widening)
    ser.write_bool(force_type_validation)
    return ser.getvalue()


def serialize_type_consistency_enforcement_qos_compact(
    kind: TypeConsistencyKind = TypeConsistencyKind.DISALLOW_TYPE_COERCION,
    ignore_sequence_bounds: bool = False,
    ignore_string_bounds: bool = False,
    ignore_member_names: bool = False,
    prevent_type_widening: bool = False,
    force_type_validation: bool = False,
    mask: int | None = None,
    endian: str = "<",
) -> bytes:
    """Serialize TypeConsistencyEnforcementQosPolicy in a compact RTI-style form.

    Encodes kind (u32) + flags mask (u32). This is not standard XCDR1 but
    matches the 8-byte length observed in RTI captures.
    """
    if mask is None:
        mask = 0
        if ignore_sequence_bounds:
            mask |= 1 << 0
        if ignore_string_bounds:
            mask |= 1 << 1
        if ignore_member_names:
            mask |= 1 << 2
        if prevent_type_widening:
            mask |= 1 << 3
        if force_type_validation:
            mask |= 1 << 4
    return struct.pack(endian + "II", int(kind), mask)

File: rust/vibedds/src/qos.rs
/// QoS policies and RxO (Requested/Offered) compatibility matching.

use crate::types::Duration;

/// Reliability QoS kind.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
#[repr(u32)]
pub enum ReliabilityKind {
    BestEffort = 1,
    Reliable = 2,
}

impl ReliabilityKind {
    pub fn from_u32(v: u32) -> Option<Self> {
        match v {
            1 => Some(Self::BestEffort),
            2 => Some(Self::Reliable),
            _ => None,
        }
    }
}

/// Durability QoS kind.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
#[repr(u32)]
pub enum DurabilityKind {
    Volatile = 0,
    TransientLocal = 1,
    Transient = 2,
    Persistent = 3,
}

impl DurabilityKind {
    pub fn from_u32(v: u32) -> Option<Self> {
        match v {
            0 => Some(Self::Volatile),
            1 => Some(Self::TransientLocal),
            2 => Some(Self::Transient),
            3 => Some(Self::Persistent),
            _ => None,
        }
    }
}

/// Ownership QoS kind.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u32)]
pub enum OwnershipKind {
    Shared = 0,
    Exclusive = 1,
}

impl OwnershipKind {
    pub fn from_u32(v: u32) -> Option<Self> {
        match v {
            0 => Some(Self::Shared),
            1 => Some(Self::Exclusive),
            _ => None,
        }
    }
}

/// Liveliness QoS kind.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
#[repr(u32)]
pub enum LivelinessKind {
    Automatic = 0,
    ManualByParticipant = 1,
    ManualByTopic = 2,
}

impl LivelinessKind {
    pub fn from_u32(v: u32) -> Option<Self> {
        match v {
            0 => Some(Self::Automatic),
            1 => Some(Self::ManualByParticipant),
            2 => Some(Self::ManualByTopic),
            _ => None,
        }
    }
}

/// Destination order QoS kind.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
#[repr(u32)]
pub enum DestinationOrderKind {
    ByReceptionTimestamp = 0,
    BySourceTimestamp = 1,
}

impl DestinationOrderKind {
    pub fn from_u32(v: u32) -> Option<Self> {
        match v {
            0 => Some(Self::ByReceptionTimestamp),
            1 => Some(Self::BySourceTimestamp),
            _ => None,
        }
    }
}

/// History QoS kind.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u32)]
pub enum HistoryKind {
    KeepLast = 0,
    KeepAll = 1,
}

impl HistoryKind {
    pub fn from_u32(v: u32) -> Option<Self> {
        match v {
            0 => Some(Self::KeepLast),
            1 => Some(Self::KeepAll),
            _ => None,
        }
    }
}

/// Aggregated QoS settings for an endpoint.
#[derive(Debug, Clone)]
pub struct QosPolicy {
    pub reliability: ReliabilityKind,
    pub reliability_max_blocking_time: Duration,
    pub durability: DurabilityKind,
    pub ownership: OwnershipKind,
    pub liveliness: LivelinessKind,
    pub liveliness_lease_duration: Duration,
    pub deadline_period: Duration,
    pub destination_order: DestinationOrderKind,
    pub history: HistoryKind,
    pub history_depth: i32,
    pub partition: Option<Vec<String>>,
}

impl Default for QosPolicy {
    fn default() -> Self {
        Self {
            reliability: ReliabilityKind::BestEffort,
            reliability_max_blocking_time: Duration::ZERO,
            durability: DurabilityKind::Volatile,
            ownership: OwnershipKind::Shared,
            liveliness: LivelinessKind::Automatic,
            liveliness_lease_duration: Duration::INFINITE,
            deadline_period: Duration::INFINITE,
            destination_order: DestinationOrderKind::ByReceptionTimestamp,
            history: HistoryKind::KeepLast,
            history_depth: 1,
            partition: None,
        }
    }
}

impl QosPolicy {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn reliable() -> Self {
        Self {
            reliability: ReliabilityKind::Reliable,
            ..Self::default()
        }
    }

    pub fn best_effort() -> Self {
        Self::default()
    }

    /// QoS for builtin endpoints (SEDP).
    pub fn builtin_endpoint() -> Self {
        Self {
            reliability: ReliabilityKind::Reliable,
            durability: DurabilityKind::TransientLocal,
            history: HistoryKind::KeepLast,
            history_depth: 1,
            ..Self::default()
        }
    }
}

/// Check if offered QoS is compatible with requested QoS (RxO rules).
///
/// Returns true if the writer's offered QoS satisfies the reader's requested QoS.
pub fn qos_compatible(offered: &QosPolicy, requested: &QosPolicy) -> bool {
    // Reliability: offered must be >= requested
    if offered.reliability < requested.reliability {
        return false;
    }

    // Durability: offered must be >= requested
    if offered.durability < requested.durability {
        return false;
    }

    // Ownership: must match exactly
    if offered.ownership != requested.ownership {
        return false;
    }

    // Liveliness: offered must be >= requested (stricter is higher)
    if offered.liveliness < requested.liveliness {
        return false;
    }

    // Destination order: offered must be >= requested
    if offered.destination_order < requested.destination_order {
        return false;
    }

    // Partition: if both specify partitions, they must have at least one in common
    if let (Some(off_parts), Some(req_parts)) = (&offered.partition, &requested.partition) {
        let has_common = off_parts.iter().any(|p| req_parts.contains(p));
        if !has_common {
            return false;
        }
    }

    true
}

/// Serialize reliability QoS for SEDP parameter list.
/// Returns kind(u32) + max_blocking_time(Duration: i32+u32) = 12 bytes
pub fn serialize_reliability_qos(qos: &QosPolicy) -> [u8; 12] {
    let mut buf = [0u8; 12];
    buf[0..4].copy_from_slice(&(qos.reliability as u32).to_le_bytes());
    buf[4..8].copy_from_slice(&qos.reliability_max_blocking_time.seconds.to_le_bytes());
    buf[8..12].copy_from_slice(&qos.reliability_max_blocking_time.fraction.to_le_bytes());
    buf
}

/// Deserialize reliability QoS from parameter value.
pub fn deserialize_reliability_qos(data: &[u8]) -> Option<(ReliabilityKind, Duration)> {
    if data.len() < 12 {
        return None;
    }
    let kind = u32::from_le_bytes([data[0], data[1], data[2], data[3]]);
    let seconds = i32::from_le_bytes([data[4], data[5], data[6], data[7]]);
    let fraction = u32::from_le_bytes([data[8], data[9], data[10], data[11]]);

    Some((ReliabilityKind::from_u32(kind)?, Duration { seconds, fraction }))
}

/// Serialize durability QoS for SEDP parameter list.
pub fn serialize_durability_qos(qos: &QosPolicy) -> [u8; 4] {
    (qos.durability as u32).to_le_bytes()
}

/// Deserialize durability QoS from parameter value.
pub fn deserialize_durability_qos(data: &[u8]) -> Option<DurabilityKind> {
    if data.len() < 4 {
        return None;
    }
    let kind = u32::from_le_bytes([data[0], data[1], data[2], data[3]]);
    DurabilityKind::from_u32(kind)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_reliability_ordering() {
        assert!(ReliabilityKind::Reliable > ReliabilityKind::BestEffort);
    }

    #[test]
    fn test_durability_ordering() {
        assert!(DurabilityKind::Persistent > DurabilityKind::TransientLocal);
        assert!(DurabilityKind::TransientLocal > DurabilityKind::Volatile);
    }

    #[test]
    fn test_qos_compatible_same() {
        let qos = QosPolicy::default();
        assert!(qos_compatible(&qos, &qos));
    }

    #[test]
    fn test_qos_compatible_reliable_to_best_effort() {
        let offered = QosPolicy::reliable();
        let requested = QosPolicy::best_effort();
        assert!(qos_compatible(&offered, &requested));
    }

    #[test]
    fn test_qos_incompatible_best_effort_to_reliable() {
        let offered = QosPolicy::best_effort();
        let requested = QosPolicy::reliable();
        assert!(!qos_compatible(&offered, &requested));
    }

    #[test]
    fn test_qos_compatible_durability() {
        let mut offered = QosPolicy::default();
        offered.durability = DurabilityKind::TransientLocal;

        let mut requested = QosPolicy::default();
        requested.durability = DurabilityKind::Volatile;

        assert!(qos_compatible(&offered, &requested));
    }

    #[test]
    fn test_qos_incompatible_durability() {
        let mut offered = QosPolicy::default();
        offered.durability = DurabilityKind::Volatile;

        let mut requested = QosPolicy::default();
        requested.durability = DurabilityKind::TransientLocal;

        assert!(!qos_compatible(&offered, &requested));
    }

    #[test]
    fn test_qos_incompatible_ownership() {
        let mut offered = QosPolicy::default();
        offered.ownership = OwnershipKind::Shared;

        let mut requested = QosPolicy::default();
        requested.ownership = OwnershipKind::Exclusive;

        assert!(!qos_compatible(&offered, &requested));
    }

    #[test]
    fn test_qos_compatible_partition_overlap() {
        let mut offered = QosPolicy::default();
        offered.partition = Some(vec!["A".to_string(), "B".to_string()]);

        let mut requested = QosPolicy::default();
        requested.partition = Some(vec!["B".to_string(), "C".to_string()]);

        assert!(qos_compatible(&offered, &requested));
    }

    #[test]
    fn test_qos_incompatible_partition_no_overlap() {
        let mut offered = QosPolicy::default();
        offered.partition = Some(vec!["A".to_string()]);

        let mut requested = QosPolicy::default();
        requested.partition = Some(vec!["B".to_string()]);

        assert!(!qos_compatible(&offered, &requested));
    }

    #[test]
    fn test_qos_compatible_one_no_partition() {
        let mut offered = QosPolicy::default();
        offered.partition = Some(vec!["A".to_string()]);

        let requested = QosPolicy::default(); // partition = None

        assert!(qos_compatible(&offered, &requested));
    }

    #[test]
    fn test_serialize_reliability() {
        let qos = QosPolicy::reliable();
        let bytes = serialize_reliability_qos(&qos);
        assert_eq!(bytes[0..4], 2u32.to_le_bytes()); // RELIABLE = 2
    }

    #[test]
    fn test_deserialize_reliability() {
        let qos = QosPolicy::reliable();
        let bytes = serialize_reliability_qos(&qos);
        let (kind, duration) = deserialize_reliability_qos(&bytes).unwrap();
        assert_eq!(kind, ReliabilityKind::Reliable);
        assert_eq!(duration, Duration::ZERO);
    }

    #[test]
    fn test_serialize_durability() {
        let mut qos = QosPolicy::default();
        qos.durability = DurabilityKind::TransientLocal;
        let bytes = serialize_durability_qos(&qos);
        assert_eq!(bytes, 1u32.to_le_bytes());
    }

    #[test]
    fn test_deserialize_durability() {
        let bytes = 2u32.to_le_bytes();
        let kind = deserialize_durability_qos(&bytes).unwrap();
        assert_eq!(kind, DurabilityKind::Transient);
    }

    #[test]
    fn test_builtin_endpoint_qos() {
        let qos = QosPolicy::builtin_endpoint();
        assert_eq!(qos.reliability, ReliabilityKind::Reliable);
        assert_eq!(qos.durability, DurabilityKind::TransientLocal);
        assert_eq!(qos.history, HistoryKind::KeepLast);
        assert_eq!(qos.history_depth, 1);
    }

    #[test]
    fn test_reliability_from_u32() {
        assert_eq!(ReliabilityKind::from_u32(1), Some(ReliabilityKind::BestEffort));
        assert_eq!(ReliabilityKind::from_u32(2), Some(ReliabilityKind::Reliable));
        assert_eq!(ReliabilityKind::from_u32(99), None);
    }

    #[test]
    fn test_durability_from_u32() {
        assert_eq!(DurabilityKind::from_u32(0), Some(DurabilityKind::Volatile));
        assert_eq!(DurabilityKind::from_u32(1), Some(DurabilityKind::TransientLocal));
        assert_eq!(DurabilityKind::from_u32(2), Some(DurabilityKind::Transient));
        assert_eq!(DurabilityKind::from_u32(3), Some(DurabilityKind::Persistent));
        assert_eq!(DurabilityKind::from_u32(99), None);
    }
}
